test/integration/consul-container/libs/consul-cluster/consul-cluster.go

```diff
@@ -0,0 +1,105 @@
+package consul_cluster
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	consulnode "github.com/hashicorp/consul/integration/consul-container/libs/consul-node"
+)
+
+// Cluster abstract a Consul Cluster by providing
+// a way to create and join a Consul Cluster
+// a way to add nodes to a cluster
+// a way to fetch the cluster leader...
+type Cluster struct {
+	Nodes []consulnode.ConsulNode
+}
+
+// New Create a new cluster based on the provided configuration
+func New(configs []consulnode.Config) (*Cluster, error) {
+	cluster := Cluster{}
+
+	for _, c := range configs {
+		n, err := consulnode.NewConsulContainer(context.Background(), c)
+		if err != nil {
+			return nil, err
+		}
+		cluster.Nodes = append(cluster.Nodes, n)
+	}
+	err := cluster.join()
+	if err != nil {
+		return nil, err
+	}
+	return &cluster, nil
+}
+
+// AddNodes add a number of nodes to the current cluster and join them to the cluster
+func (c *Cluster) AddNodes(nodes []consulnode.ConsulNode) error {
+	if len(c.Nodes) < 1 {
+		return fmt.Errorf("cannot add a node to an empty cluster")
+	}
+	n0 := c.Nodes[0]
+	for _, node := range nodes {
+		addr, _ := n0.GetAddr()
+		err := node.GetClient().Agent().Join(addr, false)
```

Comment:
nit: as we're only using `n0` to capture its address, I wonder if grabbing the address outside of the loop (and giving it a more explicit name) would help readaibility. What do you think?

```suggestion
	joinAddr, _ := c.Nodes[0].GetAddr()
	for _, node := range nodes {
		err := node.GetClient().Agent().Join(joinAddr, false)
```